modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -86,6 +86,9 @@ protected BulkRequest buildBulk(Iterable<SearchHit> docs) {
         Map<String, Object> scriptCtx = null;
 
         for (SearchHit doc : docs) {
+            if (doc.isSourceEmpty()) {
+                throw new IllegalArgumentException("[" + doc.index() + "][" + doc.type() + "][" + doc.id() + "] didn't store _source");
```

Comment:
Ok - it looks like the source can be empty if you don't request it. Since reindex lets you modify the fields parameter that could cause it. I think the right thing there is to prevent the user from modifying the fields parameter on the search request at all.